go/v1beta1/storage/elasticsearch.go

```diff
@@ -884,29 +886,95 @@ func (es *ElasticsearchStorage) genericList(ctx context.Context, log *zap.Logger
+			es.client.Search.WithSize(grafeasMaxPageSize),
+		)
+	}
+
 	encodedBody, requestJson := esutil.EncodeRequest(body)
 	log = log.With(zap.String("request", requestJson))
 	log.Debug("performing search")
 
 	res, err := es.client.Search(
-		es.client.Search.WithContext(ctx),
-		es.client.Search.WithIndex(index),
-		es.client.Search.WithBody(encodedBody),
-		es.client.Search.WithSize(grafeasMaxPageSize),
+		append(searchOptions, es.client.Search.WithBody(encodedBody))...,
 	)
 	if err != nil {
-		return nil, createError(log, "error sending request to elasticsearch", err)
+		return nil, "", createError(log, "error sending request to elasticsearch", err)
 	}
 	if res.IsError() {
-		return nil, createError(log, "unexpected response from elasticsearch", nil, zap.String("response", res.String()), zap.Int("status", res.StatusCode))
+		return nil, "", createError(log, "unexpected response from elasticsearch", nil, zap.String("response", res.String()), zap.Int("status", res.StatusCode))
 	}
 
 	var searchResults esutil.EsSearchResponse
 	if err := esutil.DecodeResponse(res.Body, &searchResults); err != nil {
-		return nil, createError(log, "error decoding elasticsearch response", err)
+		return nil, "", createError(log, "error decoding elasticsearch response", err)
+	}
+
+	return searchResults.Hits, nextPageToken, nil
+}
+
+func (es *ElasticsearchStorage) handlePagination(ctx context.Context, log *zap.Logger, body *esutil.EsSearch, index, pageToken string, pageSize int32) (string, []func(*esapi.SearchRequest), error) {
+	log = log.With(zap.String("pageToken", pageToken), zap.Int32("pageSize", pageSize))
+
+	var (
+		pit  string
+		from int
+		err  error
+	)
+
+	// if no pageToken is specified, we need to create a new PIT
+	if pageToken == "" {
+		res, err := es.client.OpenPointInTime(
+			es.client.OpenPointInTime.WithContext(ctx),
+			es.client.OpenPointInTime.WithIndex(index),
+			es.client.OpenPointInTime.WithKeepAlive(pitKeepAlive),
+		)
+		if err != nil {
+			return "", nil, createError(log, "error sending request to elasticsearch", err)
+		}
+		if res.IsError() {
+			return "", nil, createError(log, "unexpected response from elasticsearch", nil, zap.String("response", res.String()), zap.Int("status", res.StatusCode))
+		}
+
+		var pitResponse esutil.ESPitResponse
+		if err = esutil.DecodeResponse(res.Body, &pitResponse); err != nil {
+			return "", nil, createError(log, "error decoding elasticsearch response", err)
+		}
+
+		pit = pitResponse.Id
+		from = 0
+	} else {
+		// get the PIT from the provided pageToken
+		pit, from, err = esutil.ParsePageToken(pageToken)
+		if err != nil {
+			return "", nil, createError(log, "error parsing page token", err)
+		}
+	}
+
+	body.Pit = &esutil.EsSearchPit{
+		Id:        pit,
+		KeepAlive: pitKeepAlive,
 	}
 
-	return searchResults.Hits, nil
+	return esutil.CreatePageToken(pit, from+int(pageSize)), []func(*esapi.SearchRequest){
+		es.client.Search.WithSize(int(pageSize)),
```

Comment:
The documentation makes it sound like there's a limit to the number of pages when using `from`/`size`:

>By default, you cannot use from and size to page through more than 10,000 hits. This limit is a safeguard set by the index.max_result_window index setting. If you need to page through more than 10,000 hits, use the search_after parameter instead.

Does the page token need to capture the `search_after` param and encode that in the page token as well? Or is that number a reasonable cap to have?

Reply:
Regarding the 10,000 limit - I saw that as well, but since the Grafeas API layer already forbids a `pageSize` that's greater than 1000, I didn't think that this would matter.  So I don't think there's a need to use `search_after` at all, but I could be wrong there.

The index to resume searching on with the next paginated request is already encoded into the page token that's sent back to the client.

Reply:
I could be misinterpreting the docs, but it seems like the limit in Elasticsearch is on the number of search results. So if I had 10,001 results for a query, I could only retrieve 10 pages of 1000 results each and then be unable to retrieve the last page with its single result.

Reply:
Oh, I see what you mean.  Good catch, I'll have to check that out and probably test it against more than 10,000 occurrences.

I think we could probably still merge this PR to enable more of the UI work and take a look at this case in a follow up.

Reply:
For sure, I don't think it should hold up this PR (and it may not even be an issue).